README.md
# SSML Split

Splits SSML strings into batches AWS Polly ánd Google's Text to Speech API can consume.

[![Build Status](https://img.shields.io/github/workflow/status/jvandenaardweg/ssml-split/Publish%20NPM%20Package)](https://github.com/jvandenaardweg/ssml-split/actions)
[![NPM Package](https://img.shields.io/npm/dm/ssml-split.svg)](https://npmjs.com/package/ssml-split)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/0397a395ee25486395d49ababf9407d9)](https://www.codacy.com/manual/jvandenaardweg/ssml-split?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=jvandenaardweg/ssml-split&amp;utm_campaign=Badge_Grade)
[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)
[![Twitter Follow](https://img.shields.io/twitter/url/https/twitter.com/jvandenaardweg.svg?style=social&label=Follow%20%40jvandenaardweg)](https://twitter.com/jvandenaardweg)

## Features

*  Splits your large SSML into batches AWS Polly and Google's Text to Speech API can consume.
*  Makes sure you stay below the API character limitations by configuring a `hardLimit`.
*  Creates the least possible batch size to limit your requests to the Text to Speech API's.
*  Will split text at the nearest `.`, `,`, `;` or space. Can be configured.
*  Sanitizes your SSML by removing new lines, excessive white spaces, and empty tags, resulting in less characters used.
*  Uses TypeScript so you can enjoy the type safety and documentation that comes with it.

Based on [polly-ssml-split](https://github.com/oleglegun/polly-ssml-split) by [@oleglegun](https://github.com/oleglegun)

## Documentation

*  [Installation](#installation) - Walk through how to install SSML Split.
*  [Usage](#usage) - Read how SSML Split works with the available options.
*  [Recommended Options](#recommended-options) - Use these options to get started quickly.
*  [Contributing](./CONTRIBUTING.md) - Become familiar with how to contribute back to SSML Split
*  [Code of Conduct](./CODE_OF_CONDUCT.md) - Be a good citizen by following these repository rules

## Installation
Install the package with:

```sh
npm install ssml-split --save
```

## Usage

Import the package and set the options. Use the `.split()` method to split your SSML string. You can tweak the `softLimit` to see what works for you. I suggest you keep the `hardLimit` at the limitation limit of the respective API:

```typescript
import SSMLSplit from 'ssml-split';

const ssmlSplit = new SSMLSplit({
  // The service you are using: "google" or "aws"
  synthesizer: 'google',
  // Finds a possible split moment starting from 4000 characters
  softLimit: 4000,
  // Google Text to Speech limitation
  hardLimit: 5000,
  // Allow to split large paragraphs, set to false to keep your <p></p> intact
  breakParagraphsAboveHardLimit: true
});

const batches = ssmlSplit.split('<speak>your long ssml here</speak>');
```


| Option              | Type | Default                       | Description                                                                           |
| ------------------- | ---- | ------------------------- | ------------------------------------------------------------------------------------- |
| `synthesizer` | `string` | `aws`  | Set to which synthesizer you are using. Useful for when you use `breakParagraphsAboveHardLimit`. It allows the library to determine the correct break length, as that differs per synthesizer service. |
| `softLimit` | `number` | `1500`  | The amount of characters the script will start trying to break-up your SSML in multiple parts. You can tweak this number to see what works for you. |
| `hardLimit` | `number` | `3000`  | The amount of characters the script should stay below for maximum size per SSML part. If any batch size goes above this, the script will error. This hard limit is the character limit of the AWS or Google API you are using. |
| `breakParagraphsAboveHardLimit` | `boolean` | `false` | Set to `true` to allow the script to break up large paragraphs by removing the `<p>` and replacing the `</p>` with a `<break strength="x-strong" />` (for `aws`) or `<break strength="x-weak" />` (for `google`). Which results in the same pause. Requires option `synthesizer` to be set. |
| `extraSplitChars` | `string` | `,;.` | Characters that can be used as split markers for plain text.

### About: synthesizer
By using the option `synthesizer: 'google'` the library will include counting SSML tags characters to determine the best possible split moment. This makes the library also work with Google's Text to Speech API.

For example:
`<speak><p>some text</p></speak>`

The default behaviour would count that as 9 characters, which is fine for AWS Polly, but not for Google's Text to Speech API.

With `synthesizer: 'google'` it will be count as 31 characters, [just like Google's Text to Speech API counts it](https://cloud.google.com/text-to-speech/pricing?hl=en).

This should prevent you from seeing this error when using Google's Text to Speech API:

```bash
INVALID_ARGUMENT: 5000 characters limit exceeded.
```

### About: breakParagraphsAboveHardLimit
By adding the option `breakParagraphsAboveHardLimit: true` you allow the script to break up large paragraphs by removing the `<p>` and replacing the `</p>` with a `<break strength="x-strong" />` for AWS or `<break strength="x-weak" />` for Google. Which results in the same pause. This allows the library to properly split large paragraphs.

Using this option will result in 20 more characters, per paragraph, to your usage when using Google's Text to Speech API.

If you work with large paragraphs and you do not use this option, you might run into errors like `SSML tag appeared to be too long`.

Using this option is recommended when you have SSML length that goes above the `hardLimit`.

### Recommended options
#### AWS
```javascript
new SSMLSplit({
  synthesizer: 'aws',
  softLimit: 2000,
  hardLimit: 3000, // AWS Polly limitation
  breakParagraphsAboveHardLimit: true, // optional, but recommended when you have large <p>'s
})
```

#### Google
```javascript
new SSMLSplit({
  synthesizer: 'google',
  softLimit: 4000,
  hardLimit: 5000, // Google Text to Speech API limitation
  breakParagraphsAboveHardLimit: true, // optional, but recommended when you have large <p>'s
})
```

## About
The [polly-ssml-split](https://github.com/oleglegun/polly-ssml-split) by [@oleglegun](https://github.com/oleglegun) library already handles splitting of SSML correctly for AWS Polly, but wasn't working properly for Google's Text to Speech. So I just modified the package to fit my needs.

### Changes compared to `polly-ssml-split`:
*  Added `synthesizer` option to count characters based on the complete SSML tag and not just the included text characters. Which is required if you work with Google's Text to Speech API.
*  Rewrote the library to use Typescript, so you get correct type checking in your Typescript project.
*  Removed the `.configure` method and use the class constructor method for it instead.
*  Added `breakParagraphsAboveHardLimit` options to break up large paragraphs by removing the `<p>` and replacing the `</p>` with a `<break strength="x-strong" />` for AWS or `<break strength="x-weak" />` for Google. Which results in the same pause. This allows the library to properly split the paragraph and to send less batches to the text to speech API's.
*  Added more tests using Jest.

## Development

Any contribution is appreciated! Please read our [CONTRIBUTING.md](https://github.com/jvandenaardweg/ssml-split/blob/master/CONTRIBUTING.md) on how to contribute.

Use a test-driven approach when developing new features or fixing bugs.

Develop:

```bash
$ npm install
$ npm run dev
```

Run tests on file change:

```bash
$ npm test:watch
```

Build Typescript to Javascript:
```bash
$ npm run build
```

Run all tests:

```bash
$ npm test
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.5.0](https://github.com/jvandenaardweg/ssml-split/compare/0.4.1...v0.5.0)

> 8 January 2020

- Introducing new option "synthesizer" to differentiate paragraph break tags for Google and AWS and automatically count characters correctly [`#8`](https://github.com/jvandenaardweg/ssml-split/pull/8)
- Use different paragraph break tag for Google and AWS [`a5ff25c`](https://github.com/jvandenaardweg/ssml-split/commit/a5ff25c48281543f6d779089742ae14a954b18dd)
- Add some test using large text with and without punctuation [`5777659`](https://github.com/jvandenaardweg/ssml-split/commit/57776599bb3942bd0258ec3c850e8a2c10aef0fc)
- Add more tests [`33cd856`](https://github.com/jvandenaardweg/ssml-split/commit/33cd856db3c1e9e65e94b08fde3228bd397f9ede)

#### [0.4.1](https://github.com/jvandenaardweg/ssml-split/compare/0.4.0...0.4.1)

> 7 January 2020

- Remove dist dir from repo, will be included in npm package on npm [`75e02fe`](https://github.com/jvandenaardweg/ssml-split/commit/75e02feff2fb8b6d58d6feab521d4977be160dee)
- Remove options from some tests, dont need it there [`272fd6d`](https://github.com/jvandenaardweg/ssml-split/commit/272fd6da177b54814221487e1633e4ee75bce9fb)
- Release 0.4.1 [`383388d`](https://github.com/jvandenaardweg/ssml-split/commit/383388df8dfe5e8589674f46567d2a79c713cf06)

#### [0.4.0](https://github.com/jvandenaardweg/ssml-split/compare/0.3.4...0.4.0)

> 7 January 2020

- Replace double quotes with single quotes [`afa7c72`](https://github.com/jvandenaardweg/ssml-split/commit/afa7c72405e737b2ed1b898b0ca144d04efa04dc)
- Update README.md [`5444669`](https://github.com/jvandenaardweg/ssml-split/commit/5444669bebf2d5ee2218460b7e9cf775b115fe34)
- Update CONTRIBUTING.md [`fac378b`](https://github.com/jvandenaardweg/ssml-split/commit/fac378b13ecd1d81409184f2b75884044eab3636)

#### [0.3.4](https://github.com/jvandenaardweg/ssml-split/compare/0.3.3...0.3.4)

> 6 January 2020

- Update readme [`dd4c5b6`](https://github.com/jvandenaardweg/ssml-split/commit/dd4c5b604772b014a8c08e2811991e812967cbb6)
- Remove API.md [`f3b1f38`](https://github.com/jvandenaardweg/ssml-split/commit/f3b1f384d76d85cefc4e0cdc2a41630bb5f3e436)
- Release 0.3.4 [`3158136`](https://github.com/jvandenaardweg/ssml-split/commit/3158136eddbefc547bcc70041972c2bd3b7b3241)

#### [0.3.3](https://github.com/jvandenaardweg/ssml-split/compare/0.3.2...0.3.3)

> 6 January 2020

- Build result [`eda2a5b`](https://github.com/jvandenaardweg/ssml-split/commit/eda2a5b2da1d6c68d6af2b420bc383dd54b2777e)
- Use characterCounter getter [`6cbe48d`](https://github.com/jvandenaardweg/ssml-split/commit/6cbe48da35f19847d1a6b03d564450a28edfe858)
- Correctly strip SSML tags to count like AWS counts [`245eaf0`](https://github.com/jvandenaardweg/ssml-split/commit/245eaf0103ba61b385a90d0ac1ec11d76a12569d)

#### [0.3.2](https://github.com/jvandenaardweg/ssml-split/compare/0.3.1...0.3.2)

> 4 January 2020

- Release 0.3.2 [`ee27ab7`](https://github.com/jvandenaardweg/ssml-split/commit/ee27ab714cc55be9a399e70d88e888543f646a59)
- Update dist build docs [`99f4f96`](https://github.com/jvandenaardweg/ssml-split/commit/99f4f969c97b5e1d3683039e689e6983194ed832)
- Set correct default softLimit in docs [`da7c2e8`](https://github.com/jvandenaardweg/ssml-split/commit/da7c2e803b4cf3f9aaac98668c917c936f7d9cbd)

#### [0.3.1](https://github.com/jvandenaardweg/ssml-split/compare/0.3.0...0.3.1)

> 4 January 2020

- New build result [`21d917d`](https://github.com/jvandenaardweg/ssml-split/commit/21d917dcac8e52a992d96ebba1b3a58adba08720)
- Keep documentation in types on build [`f056c36`](https://github.com/jvandenaardweg/ssml-split/commit/f056c368766912364e1eb9253935cc5952fbc099)
- Add documentation to options [`c7b8deb`](https://github.com/jvandenaardweg/ssml-split/commit/c7b8debd2861dcddb80a706be251c3b25d1a7ca3)

#### [0.3.0](https://github.com/jvandenaardweg/ssml-split/compare/0.2.0...0.3.0)

> 4 January 2020

- Allow to split large paragraphs by removing &lt;p&gt;&lt;/p&gt; and adding a &lt;break /&gt; [`#7`](https://github.com/jvandenaardweg/ssml-split/pull/7)
- Replace &lt;/p&gt; with a &lt;break /&gt; to allow breakup long paragraphs [`76fff0a`](https://github.com/jvandenaardweg/ssml-split/commit/76fff0afc4c138e3f461158677b62a77671df1db)
- Exclude tests dir from build [`7b0d875`](https://github.com/jvandenaardweg/ssml-split/commit/7b0d875ec72ab564605f50c44244bdd320611c0a)
- Remove dist [`98f957e`](https://github.com/jvandenaardweg/ssml-split/commit/98f957e8b846b10db6ccb8aa53d126e089e81544)

#### [0.2.0](https://github.com/jvandenaardweg/ssml-split/compare/0.1.16...0.2.0)

> 3 January 2020

- Convert JS to Typescript [`#6`](https://github.com/jvandenaardweg/ssml-split/pull/6)
- Use Typescript [`1c04f3c`](https://github.com/jvandenaardweg/ssml-split/commit/1c04f3c1a46bcd2c18af1acccdee38f2f78708f8)
- Add latest build [`e046bd6`](https://github.com/jvandenaardweg/ssml-split/commit/e046bd6791108e4efe80c7c97f0890e1b20ca283)
- Use correct types [`1aaecf5`](https://github.com/jvandenaardweg/ssml-split/commit/1aaecf5e696e3177c5cf157808da3f4b5b4a7d7b)

#### [0.1.16](https://github.com/jvandenaardweg/ssml-split/compare/0.1.15...0.1.16)

> 2 January 2020

- Add jest tests [`#5`](https://github.com/jvandenaardweg/ssml-split/pull/5)
- Release 0.1.16 [`ccac897`](https://github.com/jvandenaardweg/ssml-split/commit/ccac897031bce8bcd17bfe3f8ffb48a53048ee3d)
- Rename test name [`b7280ca`](https://github.com/jvandenaardweg/ssml-split/commit/b7280ca5088ecf6c5512c602e5286eb49081b316)
- Run tests on push [`7567926`](https://github.com/jvandenaardweg/ssml-split/commit/7567926328d4e64f8268cdb4f22d43c2681232d2)

#### [0.1.15](https://github.com/jvandenaardweg/ssml-split/compare/0.1.14...0.1.15)

> 2 January 2020


src/index.ts
import pollyTextSplit from 'polly-text-split';
import defaults from './defaults';
import { ConfigurationValidationError, NotPossibleSplitError, SSMLParseError } from './errors';

type Synthesizer = 'google' | 'aws';

export interface OptionsInput {
  /**
   * Default: `1500`
   *
   * The amount of characters the script will start trying to break-up your SSML in multiple parts.
   * You can tweak this number to see what works for you.
   */
  softLimit?: number;
  /**
   * Default: `3000`
   *
   * The amount of characters the script should stay below for maximum size per SSML part. If any batch size goes above this, the script will error.
   */
  hardLimit?: number;
  /**
   * Default: `,;.`
   *
   * Text can be split at these given characters.
   */
  extraSplitChars?: string;
  /**
   * Default: `false`
   *
   * Set to `true` to allow the script to break up large paragraphs
   * by removing the `<p>` and replacing the `</p>` with a `<break strength="x-strong" />`,
   * which results in the same pause. Source: https://docs.aws.amazon.com/polly/latest/dg/supportedtags.html#p-tag
   *
   * This allows the script to properly split large paragraph's and to send less batches to the text to speech API's.
   *
   * It is also recommended to set this to `true` when you work with large paragraphs and experiencing errors
   * like "SSML tag appeared to be too long".
   */
  breakParagraphsAboveHardLimit?: boolean;
  /**
   * Default: `aws`
   *
   * Set to which synthesizer you are using. Useful for when you use `breakParagraphsAboveHardLimit: true`,
   * to set the correct break length, as that differs per service.
   */
  synthesizer?: Synthesizer;
}

export interface Options {
  hardLimit: number;
  softLimit: number;
  extraSplitChars: string;
  breakParagraphsAboveHardLimit: boolean;
  synthesizer: Synthesizer;
  defaultParagraphBreakSSMLTag: string;
}

interface Node  {
  parentNode: Node | null;
  type: string;
  value: string;
  children?: Node[];
}

interface RootNode extends Node {
  children: Node[];
}

/**
 * Creates a tree data structure from SSML text.
 * @class
 */
export class SSMLSplit {
  public options: Options;

  private root: RootNode;
  private batches: string[];
  private accumulatedSSML: string;
  private textLength: number;

  constructor(options?: OptionsInput) {
    if (options && typeof options !== 'object') {
      throw new ConfigurationValidationError('Parameter `options` must be an object.');
    }

    if (options && options.synthesizer && !['google', 'aws'].includes(options.synthesizer)) {
      throw new ConfigurationValidationError('Option `synthesizer` must be "google" or "aws".');
    }

    if (options && options.breakParagraphsAboveHardLimit && !options.synthesizer) {
      throw new ConfigurationValidationError('Option `synthesizer` is required when using `breakParagraphsAboveHardLimit`.');
    }

    this.setDefaults();

    this.options = {
      softLimit: options && options.softLimit || defaults.SOFT_LIMIT,
      hardLimit: options && options.hardLimit || defaults.HARD_LIMIT,
      extraSplitChars: options && options.extraSplitChars || defaults.EXTRA_SPLIT_CHARS,
      breakParagraphsAboveHardLimit: options && options.breakParagraphsAboveHardLimit || defaults.BREAK_PARAGRAPHS_ABOVE_HARD_LIMIT,
      synthesizer: options && options.synthesizer || defaults.SYNTHESIZER as Synthesizer,

      // Google and AWS apparently use different break strength pause lengths.
      // AWS uses "x-strong" as the paragraph break strength.
      // Google seems to use "x-weak"
      // So we have to differentiate between those services here.
      // AWS: https://docs.aws.amazon.com/polly/latest/dg/supportedtags.html#p-tag
      // Google: no documentation available about this topic, "x-weak" seems to add the same pause as </p>
      defaultParagraphBreakSSMLTag: options && options.synthesizer === 'google' ? defaults.GOOGLE_PARAGRAPH_BREAK_SSML_TAG : defaults.AWS_PARAGRAPH_BREAK_SSML_TAG
    };
  }

  /**
   * Split SSML text by batches of ~3000 (by default) chars.
   *
   * @throws {NotPossibleSplitError} Text cannot be split, increase `hardLimit`.
   * @throws {SSMLParseError} Argument `ssml` is not a valid SSML string.
   */
  public split(ssmlInput: string): string[] {
    // Set the defaults everytime we invoke split
    // So you can use the split method multiple times using the same class instance
    this.setDefaults();

    // Create a copy so ssmlInput stays intact when we replace paragraphs with breaks
    let ssmlToWorkWith = `${ssmlInput}`;

    if (this.options.breakParagraphsAboveHardLimit && ssmlToWorkWith.length > this.options.hardLimit) {
      // Remove paragraphs and replace it with a break.
      // This allows easier break up of long paragraphs, while maintaining the proper pause at the end.
      // Adding <break strength="" /> at the end of a paragraph is the same as wrapping your text inside a <p></p>
      ssmlToWorkWith = ssmlToWorkWith.replace(/<p>/g, '');
      ssmlToWorkWith = ssmlToWorkWith.replace(/<\/p>/g, this.options.defaultParagraphBreakSSMLTag);
    }

    // Sanitize and create tree
    this.buildTree(this.sanitize(ssmlToWorkWith));

    // Check if SSML is wrapped with <speak> tag
    if (this.root.children.length === 1 && this.root.children[0].type === 'speak') {
      // Remove global <speak> tag node
      // since the text will be split, new <speak> tags will wrap batches
      this.root.children = this.root.children[0].children!;
    }

    if (this.root.children.length === 0) {
      return this.batches;
    }

    // Start traversing root children
    this.root.children.forEach((node) => {
      // root level - can make splits here
      if (this.characterCounter < this.options.softLimit) {
        // Text node on the top level can be too long and become > SOFT_LIMIT and even HARD_LIMIT
        // So we need to explicitly check for overflow here
        if (
          node.type === 'TEXT' &&
          this.textLength + node.value.length > this.options.softLimit
        ) {
          this.splitTextNode(node);
        } else {
          // Text node is short or this is a tag node, that needs to be traversed, can't split here
          this.traverseNode(node);
        }
      } else if (this.characterCounter < this.options.hardLimit) {
        // SOFT_LIMIT is reached -> search for possible split locations
        this.makeSpeakBatch(this.accumulatedSSML);
        this.accumulatedSSML = '';
        this.textLength = 0;

        if (node.type === 'TEXT' && node.value.length > this.options.softLimit) {
          this.splitTextNode(node);
        } else {
          // Text node is short or this is a tag node, that needs to be traversed, can't split here
          this.traverseNode(node);
        }
      } else {
        throw new NotPossibleSplitError('SSML tag appeared to be too long.');
      }
    });

    if (this.characterCounter !== 0) {
      if (this.characterCounter < this.options.hardLimit) {
        this.makeSpeakBatch(this.accumulatedSSML);
      } else {
        throw new NotPossibleSplitError('Last SSML tag appeared to be too long.');
      }
    }

    return this.batches;
  }

  private get characterCounter() {
    // Since we wrap the output SSML within <speak></speak> tags,
    // we also need to include those in the calculation
    const outputSSMLLength = this.accumulatedSSML.length + '<speak></speak>'.length;

    return this.options.synthesizer === 'google' ? outputSSMLLength : this.textLength;
  }

  private setDefaults() {
    this.root = {
      parentNode: null,
      type: 'root',
      children: [],
      value: ''
    };

    this.batches = [];
    this.accumulatedSSML = '';
    this.textLength = 0;
  }

  /**
   * Sanitizes the input SSML by removing new lines, excessive white spaces, and empty tags.
   *
   * @param ssml
   */
  private sanitize(ssml: string): string {
    return ssml
    .replace(/\r?\n|\r/g, '') // remove new lines
    .replace(/>\s+</g, '><') // remove spaces between <tags>'s
    .replace(/<p>\s+<\/p>/g, '') // Remove empty <p> </p>
    .replace(/<p><\/p>/g, '') // Remove empty <p></p>
    .replace(/<s>\s+<\/s>/g, '') // Remove empty <s> </s>
    .replace(/<s><\/s>/g, '') // Remove empty <s></s>
    .replace(/<s\/>/g, '') // Remove self-closing <s/>
    .replace(/<p\/>/g, '') // Remove self-closing <p/>
    .trim(); // remove trailing and leading white space
  }

  private traverseNode(currentNode: Node): void {
    // check if node has children to check out too
    if (currentNode.children) {
      if (currentNode.type !== 'root') {
        // open tag
        this.accumulatedSSML += `<${currentNode.type}${currentNode.value}>`;
      }

      currentNode.children.forEach((node) => {
        this.traverseNode(node);
      });
      // close tag
      this.accumulatedSSML += `</${currentNode.type}>`;
    } else {
      // no children
      this.accumulatedSSML += this.noChildrenNodeToText(currentNode);
    }
  }

  private splitTextNode(node: Node): void {
    // Overflows => Text node needs to be checked for possible split location
    const localSoftLimit =
      this.textLength === 0 ? this.options.softLimit : this.options.softLimit - this.textLength;
    const localHardLimit = localSoftLimit + this.options.hardLimit - this.options.softLimit;

    pollyTextSplit.configure({
      hardLimit: localHardLimit,
      softLimit: localSoftLimit,
      extraSplitChars: this.options.extraSplitChars ? this.options.extraSplitChars : undefined,
    });

    const splitIndex = pollyTextSplit.splitIndex(node.value);

    this.makeSpeakBatch(this.accumulatedSSML + node.value.slice(0, splitIndex + 1));

    this.accumulatedSSML = node.value.slice(splitIndex + 1);
    this.textLength = this.accumulatedSSML.length;

    // If the remaining TEXT value is too long again, split it
    if (this.textLength > this.options.softLimit) {
      pollyTextSplit.configure({
        hardLimit: this.options.hardLimit,
        softLimit: this.options.softLimit,
        extraSplitChars: this.options.extraSplitChars ? this.options.extraSplitChars : undefined,
      });

      // Get text batches
      pollyTextSplit.split(this.accumulatedSSML).forEach((text: string) => {
        this.makeSpeakBatch(text);
      });

      this.accumulatedSSML = '';
      this.textLength = 0;
    }
  }

  private noChildrenNodeToText(node: Node): string {
    if (node.type === 'TEXT') {
      this.textLength += node.value.length;
      return node.value;
    } else {
      // SSML empty tag (i.e. no children)
      return `<${node.type}${node.value}/>`;
    }
  }

  /**
   * Pushes a SSML string into the batch array
   */
  private makeSpeakBatch(ssml: string): void {
    this.batches.push(`<speak>${ssml}</speak>`);
  }

  /**
   * Adds a new tree node as a parentNode child.
   */
  private addNode(parentNode: Node, newNode: Node): void {
    if (parentNode.children) {
      parentNode.children.push(newNode);
    } else {
      parentNode.children = [newNode];
    }
  }

  /**
   * Creates tree data structure from SSML text.
   */
  private buildTree(ssml: string): void {
    // remove extra space if needed
    ssml = ssml.trim();

    let text = '';
    let textHasStarted = false;
    let currentNode = this.root;

    for (let i = 0, len = ssml.length; i < len; i++) {
      // check if the char is a plain text or SSML tag
      if (ssml[i] === '<') {
        /*
         * 1. SSML tag
         */

        // check if the text was already started - finish it and add to the parentNode
        if (textHasStarted) {
          textHasStarted = false;

          const newNode: Node = {
            parentNode: currentNode,
            type: 'TEXT',
            value: text,
          };

          this.addNode(currentNode, newNode);
        }

        // type and value/attributes of parsed SSML tag
        let type = '';
        let value = ''; // can be blank, like <tag /> or <tag></tag>

        let isEndTag = false; // flag for end tag (</tag>)
        let isEmptyTag = false; // flag for empty tag (<tag />)

        // start from next char
        let j = i + 1;

        // check if it is an end tag (no value)
        if (ssml[j] === '/') {
          isEndTag = true;

          // start from the next char
          j++;

          // parse only type
          while (ssml[j] !== '>') {
            type += ssml[j];
            j++;
          }
        } else {
          /*
           *  1. Parse type unless:
           *  ' ' - value is coming
           *  '>' - is start tag marker
           *  '/' - is empty tag marker
           */
          while (ssml[j] !== ' ' && ssml[j] !== '>' && ssml[j] !== '/') {
            type += ssml[j];
            j++;
          }

          // 2. Parse value
          while (true) {
            if (ssml[j] !== '>') {
              // A. value continues -> accumulate value
              value += ssml[j];
            } else if (ssml[j - 1] === '/') {
              // B. empty tag <tag />
              isEmptyTag = true;

              // remove last `/` char from value
              if (value.length !== 0) { value = value.slice(0, value.length - 1); }
              break;
            } else {
              // C. end tag </tag>
              break;
            }
            j++;
          }
        }

        /*
         * Process parsed results
         */

        if (!isEndTag) {
          const newNode: Node = {
            parentNode: currentNode,
            type,
            value
          };
          this.addNode(currentNode, newNode);

          if (!isEmptyTag) {
            // Not an empty tag => can have other children, then keep it active
            currentNode = newNode as RootNode;
          }
        } else if (isEmptyTag) {
          // TODO: this else if might not be needed, might be removed
          // an empty tag (<break />) cannot be an end tag (</break />), doesnt make sense

          // is an end tag + empty end tag
          const newNode: Node = {
            parentNode: currentNode,
            type,
            value,
          };
          this.addNode(currentNode, newNode);
        } else {
          // is an end tag </tag>
          // close current tag = no more children

          // sanity check
          if (currentNode.type !== type) {
            const msg = `Incorrect SSML: ${type} !== ${currentNode.type}`;
            throw new SSMLParseError(msg);
          }
          currentNode = currentNode.parentNode as RootNode;
        }

        // skip processed chars for the next iteration
        i = j;
      } else {
        /*
         * 2. Plain text
         */
        if (!textHasStarted) {
          textHasStarted = true;
          text = '';
        }

        // accumulate characters
        text += ssml[i];

        if (i === len - 1 && textHasStarted) {
          // ssml ends with plain text => create node
          const newNode: Node = {
            parentNode: currentNode,
            type: 'TEXT',
            value: text,
          };

          this.addNode(currentNode, newNode);
        }
      }
    }
  }
}

export default SSMLSplit;
